yt/utilities/parallel_tools/parallel_analysis_interface.py

```diff
@@ -326,6 +326,32 @@ def barrierize(*args, **kwargs):
     return barrierize
 
 
+def parallel_root_only_then_broadcast(func):
+    """
+    This decorator blocks and calls the function on the root processor and then
+    broadcasts the results to the other processors.
+    """
+
+    @wraps(func)
+    def root_only(*args, **kwargs):
+        if not parallel_capable:
+            return func(*args, **kwargs)
+        comm = _get_comm(args)
+        if comm.rank == 0:
+            try:
+                rv0 = func(*args, **kwargs)
+            except Exception:
+                traceback.print_last()
+                rv0 = None
```

Comment:
Now that I see it with the style I suggested, I suspect this would cause `UnboundLocalError` when run in parallel (`rv0` is not defined on non-root processes). I wonder how the broadcast function works. Does it know that `None` is only a placeholder value, or does it always broadcast from root ?

Reply:
oh, right, that's why it was defined as None to begin with. 

I should be able to easily test if it causes problems. I suspect it does need to be defined -- I think it will always broadcast from root but I think the python object problem needs to be defined in some way on every process first?

Reply:
Ya, `rv0` needs to be defined or you get errors on the non-root processors. I'll switch that back in the next push. 

> I wonder how the broadcast function works. Does it know that None is only a placeholder value, or does it always broadcast from root ?

It always broadcasts from a single processor to the others: by default it assumes rank 0 is root, but you can actually specify a different rank to treat as root (i.e., the call signature for `mpi_bcast`  is `def mpi_bcast(self, data, root=0):`). But it does overwrite an existing object, so it needs to be first defined on all ranks (this is more of an MPI/mpi4py thing than it is a yt thing, check out the [examples for collective communication in mpi4py](https://mpi4py.readthedocs.io/en/stable/tutorial.html#collective-communication)).

Reply:
awesome, sorry I introduced this error !